presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ShardJoins.java

```diff
@@ -0,0 +1,212 @@
+        {
+            if (isApplicable(joinNode)) {
+                long numShards = getNumberOfShards();
+                RowExpression randomNumber = call(
+                        functionAndTypeManager,
+                        "random",
+                        BIGINT,
+                        constant((long) numShards - 1, BIGINT));
+                VariableReferenceExpression leftShardVariable = planVariableAllocator.newVariable("shard", BIGINT);
+                VariableReferenceExpression rightShardVariable = planVariableAllocator.newVariable("shard", BIGINT);
+
+                PlanNode newLeftChild = PlannerUtils.addProjections(joinNode.getLeft(), planNodeIdAllocator, planVariableAllocator, ImmutableList.of(randomNumber), ImmutableList.of(leftShardVariable));
+
+                PlanNode newRightChild = shardInput(numShards, joinNode.getRight(), rightShardVariable);
+                JoinNode.EquiJoinClause shardEquality = new JoinNode.EquiJoinClause(leftShardVariable, rightShardVariable);
+                List<JoinNode.EquiJoinClause> joinCriteria = new ArrayList<>();
+                joinCriteria.addAll(joinNode.getCriteria());
+                joinCriteria.add(shardEquality);
+                PlanNode result = new JoinNode(
+                        joinNode.getSourceLocation(),
+                        joinNode.getId(),
+                        joinNode.getStatsEquivalentPlanNode(),
+                        joinNode.getType(),
+                        newLeftChild,
+                        newRightChild,
+                        joinCriteria,
+                        joinNode.getOutputVariables(),
+                        joinNode.getFilter(),
+                        joinNode.getLeftHashVariable(),
+                        joinNode.getRightHashVariable(),
+                        joinNode.getDistributionType(),
+                        joinNode.getDynamicFilters());
+
+                return context.defaultRewrite(result);
+            }
+
+            return context.defaultRewrite(joinNode);
+        }
+
+        private boolean isApplicable(JoinNode joinNode)
+        {
+            return joinNode.getType() != FULL && joinNode.getType() != RIGHT &&
+                    (getShardedJoinStrategy(session).equals(ALWAYS) ||
+                        getShardedJoinStrategy(session).equals(COST_BASED) && shouldShardJoin(joinNode));
+        }
+
+        private boolean shouldShardJoin(JoinNode joinNode)
+        {
+            // TODO: implement based on HBO stats
+            return false;
+        }
+
+        private PlanNode shardInput(long numShards, PlanNode source, VariableReferenceExpression shardVariable)
+        {
+            checkState(numShards > 1);
+
+            Block sequence = SequenceFunction.sequenceDefaultStep(0, numShards - 1);
+            RowExpression sequenceExpression = constant(sequence, new ArrayType(BIGINT));
```

Comment:
This is using a util function in the SequenceFunction definition, this looks quite weird to me. Maybe replace it with a array[] or sequence function, it will be evaluated to constant array block during compile time.

Reply:
fixed